ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -741,6 +741,10 @@ private void processReplicaMetadataResponse(Set<MessageInfo> missingRemoteStoreM
       } else {
         // the key is present in the local store. Mark it for deletion if it is deleted in the remote store and not
         // deleted yet locally
+        // if the blob is from deprecated container, then nothing needs to be done.
+        if (skipPredicate != null && skipPredicate.test(messageInfo)) {
```

Comment:
I still can't wrap my head around the exception findKey throws. How does it happen?
If we are here at line 745, then the local key is not missing from the local store. But if we move on to findKey, it throws an exception about not able to find it? Does this mean the key is compacted between?

Reply:
I think the messageInfo is coming from replicaMetadataResponseInfo which stands for "the replica metadata response from the remote store". And during replication the deprecated containers won't be added in the missingRemoteStoreMessages list, so it will jump into line 741 no matter the local store has the key or not. And let's say the local store key has been compacted already, and for it's peer replica still have this key, then it will move on to findKey and throw exception. so in this situation, I use the skipPredicate to bypass the below procedure.

Reply:
that makes sense.